File: implementations/react/src/patterns/structural/composite/array-composite.ts
import type { JsonComponent } from "./component";
import { buildNode } from "./helpers/node-builder";
import type { Json } from "./types/json";

/**
 * Representa un nodo compuesto que maneja arreglos en una estructura JSON.
 */

export class ArrayNode implements JsonComponent {
  public keyLabel: string;
  private kids: JsonComponent[];
  constructor(keyLabel: string, arr: Json[]) {
    this.keyLabel = keyLabel;
    this.kids = arr.map((v, i) => buildNode(String(i), v));
  }
  isLeaf() { return false; }
  getPreview() { return `[ … ${this.kids.length} ]`; }
  getChildren() { return this.kids; }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/composite/array_composite.dart
import 'component.dart';

class ArrayNode implements JsonComponent {
  @override
  final String keyLabel;
  final List<JsonComponent> _kids;

  ArrayNode(this.keyLabel, List<dynamic> arr, JsonComponent Function(String, dynamic) buildNode)
      : _kids = List.generate(arr.length, (i) => buildNode(i.toString(), arr[i]));

  @override
  bool isLeaf() => false;

  @override
  String getPreview() => '[ … ${_kids.length} ]';

  @override
  List<JsonComponent> getChildren() => _kids;
}
